rclone/rclone
# backend: Add NeoFS backend

### What is the purpose of this change?

Adding support for NeoFS as a backend.

[NeoFS](https://fs.neo.org/) is a distributed, decentralized object storage network developed by [Neo SPCC](https://nspcc.ru/en). Built with Neo blockchain integration in mind, NeoFS aims to support the shift away from third-party storage providers, providing users with complete control over their data.

NeoFS is backet/container based backend that support the following operations:

* Create/get/delete/getEACL/setEACL container - see appropriate [grpc service](https://github.com/nspcc-dev/neofs-api/blob/d95228c40283cf6e188073a87a802af7e5dc0a7d/container/service.proto#L17) (EACL is a table with right access rules)
* Put/get/delete/head/range object - see appropriate [grpc service](https://github.com/nspcc-dev/neofs-api/blob/d95228c40283cf6e188073a87a802af7e5dc0a7d/object/service.proto#L14)

More detail you can find in [spec](https://nspcc.ru/upload/neofs-spec-latest.pdf).

### How to test this new backend

Run NeoFS using devenv (see README in [repo](https://github.com/nspcc-dev)). You can also use [testnet, mainnet](https://docs.neo.org/docs/en-us/develop/network/testnet.html) 

Prepare rclone config. The following config for devenv: 
```
[TestNeoFS]
type = neofs
neofs_endpoint = s01.neofs.devenv:8080
wallet = ./github/nspcc/neofs-dev-env/wallets/wallet.json
basic_acl = public-read-write
placement_policy = REP 3
rpc_address = http://morph-chain.neofs.devenv:30333
```

 Now you can test new neofs backend.

### Additional links
* https://github.com/nspcc-dev/ 
* https://neo.org/ 
* https://neospcc.medium.com/neofs-and-storj-comparison-efa426504cbb 


#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

#### How to use GitHub

Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.

Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.

Subscribe to receive notifications on status change and new comments.
